internal/venice-common/src/main/java/com/linkedin/venice/pubsub/adapter/kafka/TopicPartitionsOffsetsTracker.java

```diff
@@ -64,51 +55,81 @@ public TopicPartitionsOffsetsTracker(@Nonnull Duration offsetsUpdateInterval, Du
+       *       We do allow an interval of zero in tests, via the package-private constructor, because these are not
+       *       multi-threaded.
+       */
+      if (lastMetricsCollectedTimeSnapshot == lastMetricsCollectedTime) {
+        lastMetricsCollectedTime = currentTime;
+      } else {
+        return; // Another thread got in first.
+      }
+    }
 
     // Update current offset cache for all topics partition.
     List<ConsumerRecord<byte[], byte[]>> listOfRecordsForOnePartition;
     ConsumerRecord<byte[], byte[]> lastConsumerRecordOfPartition;
     for (TopicPartition tp: records.partitions()) {
       listOfRecordsForOnePartition = records.records(tp);
       lastConsumerRecordOfPartition = listOfRecordsForOnePartition.get(listOfRecordsForOnePartition.size() - 1);
-      topicPartitionCurrentOffset.put(tp, (double) lastConsumerRecordOfPartition.offset());
+      topicPartitionOffsetInfo.computeIfAbsent(tp, k -> new OffsetInfo()).currentOffset =
+          lastConsumerRecordOfPartition.offset();
     }
 
     MetricName metricName;
     Metric metric;
     TopicPartition tp;
-    Double currOffset;
-    for (Map.Entry<MetricName, ? extends Metric> entry: metrics.entrySet()) {
+    OffsetInfo offsetInfo;
+    long metricValue;
+    for (Map.Entry<MetricName, ? extends Metric> entry: kafkaConsumer.metrics().entrySet()) {
       metricName = entry.getKey();
       metric = entry.getValue();
+      metricValue = getMetricEntryRecordsLag(metricName, metric);
 
-      if (isMetricEntryRecordsLag(metricName, metric)) {
+      if (metricValue != INVALID) {
         tp = new TopicPartition(metricName.tags().get("topic"), Integer.parseInt(metricName.tags().get("partition")));
-        currOffset = topicPartitionCurrentOffset.get(tp);
-        if (currOffset != null) {
-          topicPartitionEndOffset.put(tp, currOffset + ((Double) metric.metricValue()));
+        offsetInfo = topicPartitionOffsetInfo.computeIfAbsent(tp, k -> new OffsetInfo());
+        offsetInfo.lag = metricValue;
+        if (offsetInfo.currentOffset != INVALID) {
+          offsetInfo.endOffset = offsetInfo.currentOffset + metricValue;
         }
       }
     }
   }
 
-  private boolean isMetricEntryRecordsLag(MetricName metricName, Metric metric) {
+  /**
+   * @return the lag value if the metric is for lag, or {@link #INVALID} otherwise.
+   */
+  private long getMetricEntryRecordsLag(MetricName metricName, Metric metric) {
     try {
-      return Objects.equals(metricName.name(), "records-lag") && (metric.metricValue() instanceof Double);
+      if (Objects.equals(metricName.name(), "records-lag")) {
+        Object metricValue = metric.metricValue();
+        if (metricValue instanceof Double) {
+          // Double is the way all metrics are internally represented in Kafka, but since we are dealing with lag, we
+          // want an integral type, so we cast it.
+          return ((Double) metricValue).longValue();
+        }
+      }
+      return INVALID;
```

Comment:
Since INVALID returned for all cases except when metricValue is Double we can put it in the end of function and have single `return INVALID`?